File: packages/node/src/sdk/validation.ts
import Ajv2020, { ErrorObject, ValidateFunction } from 'ajv/dist/2020';
import addFormats from 'ajv-formats';

import { ValidationErrorDetail, ValidationMode, ValidationResult } from '../api/types';
import logEventSchema from './log-event.schema.json';

/**
 * Raised when a log event fails canonical schema validation in STRICT mode.
 */
export class ValidationError extends Error {
  /**
   * Create an error describing sanitized schema-rule failures.
   *
   * @param errors - Schema-rule errors that do not contain caller-supplied values.
   */
  constructor(errors: ValidationErrorDetail[]) {
    super(`Log event failed schema validation: ${ValidationError.formatErrors(errors)}`);
    this.name = 'ValidationError';
  }

  private static formatErrors(errors: ValidationErrorDetail[]): string {
    return errors.map(({ field, message, rule }) => `${field}: ${message} (${rule})`).join('; ');
  }
}

/**
 * Validate structured log events against the build-derived canonical schema.
 */
export class LogEventValidator {
  private readonly validator: ValidateFunction;

  /**
   * Load and compile the canonical DT3 log-event validation schema artifact.
   */
  constructor() {
    const ajv = new Ajv2020({ allErrors: true, strict: false });
    addFormats(ajv);
    this.validator = ajv.compile(logEventSchema);
  }

  // PUBLIC_INTERFACE
  /**
   * Validate an event without mutating the supplied value.
   *
   * @param event - Structured log event to validate.
   * @param mode - Repository-defined validation mode.
   * @returns The validation outcome with structured, sanitized schema-rule diagnostics.
   * @throws Error if the validation mode is not repository-defined.
   */
  public validate(
    event: Readonly<Record<string, unknown>>,
    mode: ValidationMode | string = ValidationMode.LENIENT,
  ): ValidationResult {
    const normalizedMode = String(mode).toUpperCase() as ValidationMode;

    if (!Object.values(ValidationMode).includes(normalizedMode)) {
      throw new Error('validation.mode must be one of STRICT, LENIENT, or OFF');
    }

    if (normalizedMode === ValidationMode.OFF) {
      return { valid: true, errors: [], mode: normalizedMode };
    }

    const valid = this.validator({ ...event }) as boolean;
    const errors = valid
      ? []
      : [...(this.validator.errors ?? [])]
          .sort((left, right) => this.errorSortKey(left).localeCompare(this.errorSortKey(right)))
          .map((error) => this.formatError(error));

    return { valid, errors, mode: normalizedMode };
  }

  private errorSortKey(error: ErrorObject): string {
    return `${error.instancePath}:${error.keyword}`;
  }

  private formatError(error: ErrorObject): ValidationErrorDetail {
    if (error.keyword === 'required') {
      const missingProperty = (error.params as { missingProperty?: string }).missingProperty ?? 'unknown';
      return {
        field: missingProperty,
        message: 'Required property is missing.',
        rule: 'required',
      };
    }

    return {
      field: this.toFieldPath(error.instancePath),
      message: this.messageForRule(error.keyword),
      rule: error.keyword,
    };
  }

  private toFieldPath(instancePath: string): string {
    if (!instancePath) {
      return '$';
    }

    return instancePath
      .slice(1)
      .split('/')
      .map((segment) => segment.replace(/~1/g, '/').replace(/~0/g, '~'))
      .join('.');
  }

  private messageForRule(rule: string): string {
    const messages: Record<string, string> = {
      enum: 'Value must be one of the allowed options.',
      format: 'Value does not match the required format.',
      minLength: 'Value is shorter than the minimum permitted length.',
      minimum: 'Value is below the minimum permitted value.',
      pattern: 'Value does not match the required pattern.',
      type: 'Value has an invalid type.',
    };

    return messages[rule] ?? `Value violates the '${rule}' schema rule.`;
  }
}

File: packages/python/dt3_sdk/validation.py
"""Runtime validation for canonical DT3 structured log events."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Mapping

from jsonschema import Draft202012Validator, FormatChecker


class ValidationError(ValueError):
    """Raised when a log event fails validation in STRICT mode."""


@dataclass(frozen=True)
class ValidationResult:
    """The sanitized outcome of validating one log event."""

    valid: bool
    errors: List[str]
    mode: str


class LogEventValidator:
    """Validate log events using the repository's canonical JSON Schema."""

    def __init__(self) -> None:
        """Load and compile the canonical ``log-event.schema.json`` schema."""
        schema_path = Path(__file__).resolve().parents[3] / "schemas" / "log-event.schema.json"
        with schema_path.open(encoding="utf-8") as schema_file:
            schema = json.load(schema_file)

        self._validator = Draft202012Validator(schema, format_checker=FormatChecker())

    # PUBLIC_INTERFACE
    def validate(self, event: Mapping[str, Any], mode: str = "LENIENT") -> ValidationResult:
        """Validate a structured log event against the canonical repository schema.

        Args:
            event: Event fields to validate. The mapping is never mutated.
            mode: The configured validation mode: ``STRICT``, ``LENIENT``, or ``OFF``.

        Returns:
            A validation result containing only schema-rule details, never input values.

        Raises:
            ValueError: If ``mode`` is not one of the repository-defined modes.
        """
        normalized_mode = mode.upper()
        if normalized_mode not in {"STRICT", "LENIENT", "OFF"}:
            raise ValueError(
                "validation.mode must be one of STRICT, LENIENT, or OFF"
            )

        if normalized_mode == "OFF":
            return ValidationResult(valid=True, errors=[], mode=normalized_mode)

        errors = [
            self._format_error(error)
            for error in sorted(self._validator.iter_errors(dict(event)), key=self._error_sort_key)
        ]
        return ValidationResult(
            valid=not errors,
            errors=errors,
            mode=normalized_mode,
        )

    @staticmethod
    def _error_sort_key(error: Any) -> tuple[str, str]:
        """Produce stable ordering for JSON Schema errors."""
        path = ".".join(str(part) for part in error.absolute_path)
        return path, error.validator

    @staticmethod
    def _format_error(error: Any) -> str:
        """Format an error without including caller-supplied values."""
        if error.validator == "required":
            # jsonschema exposes the complete required-field list in
            # ``validator_value``. Its message identifies the one field that is
            # actually absent, which is the only schema detail we should report.
            missing_field = error.message.split("'", 2)[1]
            return f"required property is missing ({missing_field})"

        field = ".".join(str(part) for part in error.absolute_path) or "$"
        return f"{field}: violates schema rule '{error.validator}'"
